File: 2024/src/03/day03b.m

fileID = fopen('input/day03.in', 'r');
pattern = 'mul\((\d+),(\d+)\)|(do\(\))|(don\''t\(\))';

result = 0;
readInstruction = true;
while ~feof(fileID)
    line = fgetl(fileID);
    tokens = regexp(line, pattern, 'tokens');
    if ~isempty(tokens)
        for i = 1:length(tokens)
            token = tokens{i};
            if strcmp(token{1},'do()')
                readInstruction = true;
            elseif strcmp(token{1},'don''t()')
                readInstruction = false;
            else
                if readInstruction
                    num1 = str2double(token{1});
                    num2 = str2double(token{2});
                    result = result + num1 * num2;
                end
            end
        end
    end
end
fprintf('Result: %d\n', result);
fclose(fileID);

File: 2024/src/03/day03b.cc
#include <bits/stdc++.h>
using namespace std;

int main()
{
    int result{0};
    bool readInstruction{true};
    ifstream file{"../input/day03.in"};
    string line;
    regex re(R"(mul\((\d+),(\d+)\)|do\(\)|don't\(\))");
    while (getline(file, line))
    {
        for (sregex_iterator it(line.begin(), line.end(), re), end_it; it != end_it; ++it)
        {
            smatch match = *it;
            if (match[0] == "do()")
                readInstruction = true;
            else if (match[0] == "don't()")
                readInstruction = false;
            else
            {
                if (readInstruction)
                {
                    int a = stoi(match[1]);
                    int b = stoi(match[2]);
                    result += a * b;
                }
            }
        }
    }
    cout << result << endl;
}